File: DSA/Queue/python/queue.py
#!/bin/env python3

"""
Implementation of Queue (FIFO).
"""

class Queue:
    """
    A class representing a basic Queue with fixed capacity.
    """

    def __init__(self, capacity):
        """Initialize the queue."""
        self.capacity = capacity
        self.data = []
        self.head = -1
        self.tail = -1

    def is_empty(self):
        """Check if the queue is empty."""
        return len(self.data) == 0

    def is_full(self):
        """Check if the queue is full."""
        return len(self.data) == self.capacity

    def enqueue(self, item):
        """Add an item to the queue."""
        if self.is_full():
            print("This queue is full. Cannot add item.")
            return -1
        if self.is_empty():
            self.head = 0
        self.tail += 1
        self.data.append(item)
        print(f"Item {item} added successfully to the queue.")
        return item

    def dequeue(self):
        """Remove an item from the front of the queue."""
        if self.is_empty():
            print("This queue is empty. Nothing to dequeue.")
            return -1
        item_dequeued = self.data.pop(0)
        self.tail -= 1
        print(f"Front item {item_dequeued} dequeued successfully.")
        if self.is_empty():
            self.head = -1
            self.tail = -1
        return item_dequeued

    def peek(self):
        """Return the front item without removing it."""
        if self.is_empty():
            print("This queue is empty.")
            return None
        return self.data[0]

    def size(self):
        """Return the current number of items in the queue."""
        return len(self.data)

    def show(self):
        """Display the current queue."""
        if self.is_empty():
            print("This queue is empty. Nothing to show.")
            return []
        return self.data

File: DSA/Queue/javascript/queue.js
#!/bin/env node

/**
 * Implementation of Queue (FIFO).
 */

class Queue {
  /**
   * A class representing a basic Queue with fixed capacity.
   */

  /**
   * Constructor Function to Initialize a new Queue.
   * @param {*} capacity - The Fixed size of Queue.
   */
  constructor(capacity) {
    this.capacity = capacity;
    this.data = [];
    this.head = -1;
    this.tail = -1;
  }

  /**
   * Check if the queue is empty.
   */
  isEmpty() {
    return this.data.length === 0;
  }

  /**
   * Check if the queue is full.
   */
  isFull() {
    return this.data.length === this.capacity;
  }

  /**
   * Adding a new item to the Queue (Last In).
   * @param {*} item - The Item to be Enqueued.
   */
  enqueue(item) {
    if (this.isFull()) {
      console.log("This queue is full. Cannot add item.");
      return null;
    }

    if (this.isEmpty()) {
      this.head = 0;
    }
    this.tail += 1;
    this.data.push(item);
    console.log(`Item ${item} added successfully to the queue.`);
    return item;
  }

  /**
   * Remove an item from the front of the queue.
   */
  dequeue() {
    if (this.isEmpty()) {
      console.log("This queue is empty. Nothing to dequeue.");
      return null;
    }
    let itemToDequeue = this.data.shift();
    this.tail -= 1;
    console.log(`Front item ${itemToDequeue} dequeued successfully.`);
    if (this.isEmpty()) {
      this.head = -1;
      this.tail = -1;
    }
    return itemToDequeue;
  }

  /**
   * Return the front item without removing it.
   */
  peek() {
    if (this.isEmpty()) {
      console.log("This queue is empty.");
      return null;
    }
    return this.data[0];
  }

  /**
   * Return the current number of items in the queue.
   */
  size() {
    return this.data.length;
  }

  /**
   * Display the current queue.
   */
  show() {
    if (this.isEmpty()) {
      console.log("This queue is empty. Nothing to show.");
      return [];
    }
    return this.data;
  }
}
